src/Features/Core/Portable/Completion/Providers/AbstractRecommendationServiceBasedCompletionProvider.cs

```diff
@@ -22,12 +22,15 @@ internal abstract class AbstractRecommendationServiceBasedCompletionProvider<TSy
         where TSyntaxContext : SyntaxContext
     {
         protected abstract Task<bool> ShouldPreselectInferredTypesAsync(CompletionContext? completionContext, int position, CompletionOptions options, CancellationToken cancellationToken);
-        protected abstract CompletionItemRules GetCompletionItemRules(ImmutableArray<(ISymbol symbol, bool preselect)> symbols, TSyntaxContext context);
+        protected abstract CompletionItemRules GetCompletionItemRules(ImmutableArray<SymbolAndSelectionInfo> symbols, TSyntaxContext context);
         protected abstract CompletionItemSelectionBehavior PreselectedItemSelectionBehavior { get; }
         protected abstract bool IsInstrinsic(ISymbol symbol);
         protected abstract bool IsTriggerOnDot(SyntaxToken token, int characterPosition);
 
-        protected sealed override async Task<ImmutableArray<(ISymbol symbol, bool preselect)>> GetSymbolsAsync(
+        protected virtual string GetFilterText(ISymbol symbol, string displayText, TSyntaxContext context)
+            => GetFilterTextDefault(symbol, displayText, context);
```

Comment:
this only has two subclasses right?  if so, consider not having a virtual fro just one to override.  just make abstract and have each subclass be clear in what it does for this.